src/stan_math_backend/Transform_Mir.ml

```diff
@@ -395,6 +395,65 @@ let%expect_test "insert before" =
   [%sexp (l : int list)] |> print_s ;
   [%expect {| (1 2 3 4 5 999 6) |}]
 
+let rec add_validate_dims outvars stmts =
+  let params =
+    List.filter_map
+      ~f:(function
+        | decl_id, Program.({out_block= Parameters; out_trans; _}) -> (
+          match out_trans with
+          | Program.Simplex ->
+              Some (decl_id, Internal_fun.FnValidateSizeSimplex)
+          | Program.UnitVector -> Some (decl_id, FnValidateSizeUnitVector)
+          | _ -> None )
+        | _ -> None)
+      outvars
+    |> String.Map.of_alist_exn
+  in
+  let fnvalidate id =
+    Option.value ~default:Internal_fun.FnValidateSize (Map.find params id)
+  in
+  let f = function
+    | Stmt.Fixed.({pattern= Decl {decl_id; decl_type= Sized st; _}; meta}) as
+      decl ->
+        let check fn x =
+          Stmt.Fixed.
+            { meta
+            ; pattern=
+                NRFunApp
+                  ( CompilerInternal
+                  , Internal_fun.to_string fn
+                  , Expr.Helpers.
+                      [ str decl_id
+                      ; str (Fmt.strf "%a" Expression_gen.pp_expr x)
+                      ; x ] ) }
+        in
+        let rec map = function
+          | [] -> []
+          | [s] -> [check (fnvalidate decl_id) s]
+          | s :: ls -> check FnValidateSize s :: map ls
+        in
+        map (SizedType.get_dims st) @ [decl]
+    | stmt -> [validate_dims_stmt stmt]
+  in
+  List.concat_map ~f stmts
+
+and validate_dims_stmt stmt =
+  Stmt.Fixed.
+    { stmt with
+      pattern=
+        ( match stmt.pattern with
```

Comment:
I think this would be more clearly written as 

```ocaml
let pattern = match stmt.pattern with
 | Stmt.Fixed.Pattern.Block s ->
￼            Stmt.Fixed.Pattern.Block (add_validate_dims [] s)
￼        | SList s -> SList (add_validate_dims [] s)
￼        | While (a, b) -> While (a, validate_dims_stmt b)
￼        | For f -> For {f with body= validate_dims_stmt f.body}
￼        | IfElse (p, t, e) ->
￼            IfElse
￼              ( p
￼              , validate_dims_stmt t
￼              , Option.map ~f:validate_dims_stmt e )
￼        | s -> s
in
Stmt.Fixed.{stmt with pattern}
```